README.md
# DatelessTime

A class to handle __dateless time values__.  

## What

There are a number of use cases where we need a simple and dumb time value. Think about the opening times of a theater or the daily airing time of a TV show.  

In these situations the time doesn't need to be bound to a date.  
It doesn't need to be adjusted to the user's timezone either, and it surely doesn't need to be pushed forward or backwards because of the dailight saving time.

Things become even messier when working with a framework (Rails) that handles these adjustments automatically.  

This gem aims to fix this.  
Drawing inspirations from the Unix time format, it stores static time values as the number of seconds since midnight.  
It generates static and date-independent time values, that don't care about timezones or DST.

As easy as pie.


## Dependencies

No external gems.  
`date` from the stdlib.  


## Rubies

Tested with:

* MRI `>= 1.9.3`
* Rubinius `>= 2.2`
* JRuby `>= 1.7`


## How


Creation:

```ruby
require 'dateless_time'

# you can create a DatelessTime object with a shortcut
time = DatelessTime.now

# or a Time object
time = DatelessTime.new Time.now

# or a DateTime object
time = DatelessTime.new DateTime.now

# or a String (minutes and seconds are optional)
time = DatelessTime.new "13:37:00"

# or the number of seconds since midnight
time = DatelessTime.new 49020

# or a Hash (minutes and seconds are optional)
time = DatelessTime.new hours: 13, minutes: 37, seconds: 0

# or an Array (minutes and seconds are optional)
time = DatelessTime.new [13, 37, 0]

# or another DatelessTime object
time = DatelessTime.new DatelessTime.new('00:42')

```

Interface:

```ruby
require 'dateless_time'

time = DatelessTime.new "13:37:42"

time.hours
time.hour
#=> 13

time.minutes
time.min
#=> 37

time.seconds
time.sec
#=> 42

time.seconds_since_midnight
time.to_i
# => 49062

# this uses Time.now to fill the date-related bits
time.to_time
# => 2014-04-01 13:37:42 +0100

# but you can supply a base Time object instead
time.to_time(Time.new(1985, 10, 25, 0, 0, 0, "-08:00"))
# => 1985-10-25 13:37:42 -0800

# or a base Date or DateTime
time.to_time(Date.new(1985, 10, 25))
# => 1985-10-25 13:37:42 +0100

time.to_time.class
# => Time

# this uses DateTime.now to fill the date-related bits
time.to_datetime
# => #<DateTime: 2015-11-11T13:37:42+00:00 ((2457338j,49062s,0n),+0s,2299161j)>

# but here, too, you can supply a base object instead (Date, Time or DateTime)
time.to_datetime(Date.new(1985, 10, 25))
# => #<DateTime: 1985-10-25T13:37:42+00:00 ((2446364j,49062s,0n),+0s,2299161j)>

time.strftime("%-l:%M %P")
# => "1:37 pm"

time.to_s
# => "13:37:42"

time.to_h
# => {:hours=>13, :minutes=>37, :seconds=>42}

time.to_a
# => [13, 37, 42]

```

Comparisons:

```ruby

@time_1 = DatelessTime.new "11:22"
@time_2 = DatelessTime.new "11:30"

# all the usual suspects

@time_1 < @time_2
# => true

@time_1 == @time_2
# => false

@time_2.between? @time_1, DatelessTime.new("20:30")
# => true

array = (1..5).map { DatelessTime.new(rand(DatelessTime::SECONDS_IN_24_HOURS)) }
array.map(&:to_s)
# => ["06:51:58", "04:50:32", "14:36:53", "21:36:38", "10:17:12"]
array.sort.map(&:to_s)
# => ["04:50:32", "06:51:58", "10:17:12", "14:36:53", "21:36:38"]

# etc...

```

## Installation

Add this line to your application's Gemfile:

    gem 'dateless_time'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install dateless_time



## Integration with Rails

The main goal is to keep this gem as small and lightweight as possible, thus I'm not planning to add any specific support for rails.  

This doesn't mean that it can't be used with Rais, though!  
Just choose how to store time values in you DB (time-only SQL values or seconds since midnight, for example), and use them to instantiate `DatelessTime` objects rather than Ruby's default `Time`.

For example:

```ruby

# opening_time_seconds is an INT value from the DB

def opening_time
  @opening_time ||= DatelessTime.new opening_time_seconds
end

```


## To Do


1. implement the `+` and `-` artimetic operators, in a way consistent with Ruby's `Time`
2. nice to have: other methods from `Time`'s public interface
3. using `DateTime` objects as base for `#to_time` and `#to_datetime` does not support fractional utc offsets (e.g. Australia)





## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request


CHANGELOG.md
# DatelessTime Changelog

## 1.0.0

The gem graduates from `0.0.x`. The basic functionality is stable and not going to change. The other additions that go into this release are small incremental changes.

#### New Features

* Added a changelog.  
* The method `#to_time` now also accepts `DateTime` objects as base values, in addition to `Time` and `Date`.  
* Added `#to_datetime` method, with semantics similar to `#to_time`.

#### Changes

* Small internal improvements.

## 0.0.4
#### New Features

* Added method aliases for `hour`, `min` and `sec`, to make the interface more consistent with the default `Time` class.  
* The method `#to_time` now also accepts `Date` objects as base values, in addition to `Time` objects.

## 0.0.3
#### New Features

* Added support to initialize new `DatelessTime` objects with other existing `DatelessTime` objects as sources.  
* Added support to initizlize new `DatelessTime` objects with `DateTime` objects as sources.  
* Offical support for sorting in arrays (just added tests to make sure it works).  

#### Changes

* The gem now explicitly requires `date` from the standard library.

#### Bug Fixes

* Inizializing `DatelessTime` objects with strings like `"12:01 am"` or `"12:01 pm"` was generating invalid data.

## 0.0.2

#### New Features

* Implemented support for the `Comparable` module. `DatelessTime` objects now support comparison operators: `==`, `<`, etc.

#### Bug Fixes

* The return value of `#to_time` was memoized, this made it impossible to use it multiple times with different base values.

## 0.0.1

The initial release.


test/query_test.rb
require 'test_helper'

class DateTime
  def utc_offset  #patch DateTime
    self.to_time.utc_offset 
  end
end
class QueryTest < Minitest::Test

  def setup
    @dl_time = DatelessTime.new [13, 37, 42]
  end


=begin
Test to_time
=end

  def test_to_time_without_base
    @to_time = @dl_time.to_time
    now = Time.now
    expected = Time.new(now.year, now.month, now.day, 13, 37, 42, now.utc_offset)

    assert_equal Time, @to_time.class
    assert_equal expected, @to_time
  end


  def test_to_time_with_time_base
    base = Time.new(1990, 11, 10, 12, 13, 14)
    @to_time = @dl_time.to_time(base)
    
    assert_equal Time, @to_time.class
    assert_equal Time.new(1990, 11, 10, 13, 37, 42), @to_time
  end


  def test_to_time_with_time_with_offset_base
    base = Time.new(1990, 11, 10, 12, 13, 14, '+04:00')
    @to_time = @dl_time.to_time(base)
    
    assert_equal Time, @to_time.class
    assert_equal Time.new(1990, 11, 10, 13, 37, 42, '+04:00'), @to_time
  end


  def test_to_time_with_date_base
    base = Date.new(1990, 11, 10)
    @to_time = @dl_time.to_time(base)
    
    assert_equal Time, @to_time.class
    assert_equal Time.new(1990, 11, 10, 13, 37, 42), @to_time
  end



  def test_to_time_with_datetime_base
    base = DateTime.new(1990, 11, 10, 12, 13, 14)
    @to_time = @dl_time.to_time(base)
    
    assert_equal Time, @to_time.class
    assert_equal Time.new(1990, 11, 10, 13, 37, 42), @to_time
  end


  def test_to_time_with_datetime_with_offset_base
    base = DateTime.new(1990, 11, 10, 12, 13, 14, '+4')
    @to_time = @dl_time.to_time(base)
    
    assert_equal Time, @to_time.class
    assert_equal Time.new(1990, 11, 10, 13, 37, 42, '+04:00'), @to_time
  end


  def test_to_time_with_different_bases_should_change
    @to_time = @dl_time.to_time
    now = Time.now
    expected = Time.new(now.year, now.month, now.day, 13, 37, 42, now.utc_offset)

    assert_equal Time, @to_time.class
    assert_equal expected, @to_time


    base = Time.new(1990, 11, 10, 12, 13, 14)
    @to_time = @dl_time.to_time(base)
    
    assert_equal Time, @to_time.class
    assert_equal Time.new(1990, 11, 10, 13, 37, 42), @to_time
  end

=begin
Test to_datetime
=end

  def test_to_datetime_without_base
    @to_datetime = @dl_time.to_datetime
    now = DateTime.now
    expected = DateTime.new(now.year, now.month, now.day, 13, 37, 42, now.utc_offset)

    assert_equal DateTime, @to_datetime.class
    assert_equal expected, @to_datetime
  end


  def test_to_datetime_with_time_base
    base = DateTime.new(1990, 11, 10, 12, 13, 14)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end


  def test_to_datetime_with_date_base
    base = Date.new(1990, 11, 10)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end


  def test_to_datetime_with_different_bases_should_change
    @to_datetime = @dl_time.to_datetime
    now = DateTime.now
    expected = DateTime.new(now.year, now.month, now.day, 13, 37, 42, now.utc_offset)

    assert_equal DateTime, @to_datetime.class
    assert_equal expected, @to_datetime


    base = DateTime.new(1990, 11, 10, 12, 13, 14)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end

=begin
=end






  def test_to_datetime_without_base
    @to_datetime = @dl_time.to_datetime
    now = DateTime.now
    expected = DateTime.new(now.year, now.month, now.day, 13, 37, 42)

    assert_equal DateTime, @to_datetime.class
    assert_equal expected, @to_datetime
  end


  def test_to_datetime_with_time_base
    base = Time.new(1990, 11, 10, 12, 13, 14)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end


  def test_to_datetime_with_time_with_offset_base
    base = Time.new(1990, 11, 10, 12, 13, 14, '+04:00')
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42, '+4'), @to_datetime
  end


  def test_to_datetime_with_date_base
    base = Date.new(1990, 11, 10)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end



  def test_to_datetime_with_datetime_base
    base = DateTime.new(1990, 11, 10, 12, 13, 14)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end


  def test_to_datetime_with_datetime_with_offset_base
    base = DateTime.new(1990, 11, 10, 12, 13, 14, '+4')
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42, '+4'), @to_datetime
  end


  def test_to_datetime_with_different_bases_should_change
    @to_datetime = @dl_time.to_datetime
    now = Time.now
    expected = DateTime.new(now.year, now.month, now.day, 13, 37, 42)

    assert_equal DateTime, @to_datetime.class
    assert_equal expected, @to_datetime


    base = Time.new(1990, 11, 10, 12, 13, 14)
    @to_datetime = @dl_time.to_datetime(base)
    
    assert_equal DateTime, @to_datetime.class
    assert_equal DateTime.new(1990, 11, 10, 13, 37, 42), @to_datetime
  end





  def test_to_s
    assert_equal "13:37:42", @dl_time.to_s
  end


  def test_to_s_with_padding_zeroes
    @dl_time = DatelessTime.new [2, 7, 9]
    assert_equal "02:07:09", @dl_time.to_s
  end


  def test_to_h
    assert_equal({ hours: 13, minutes: 37, seconds: 42 }, @dl_time.to_h)
  end


  def test_to_a
    assert_equal [13, 37, 42], @dl_time.to_a
  end


  def test_seconds_since_midnight
    secs = 42 + (37 * 60) + (13 * 60 * 60)
    assert_equal secs, @dl_time.seconds_since_midnight
  end


  def test_to_i
    secs = 42 + (37 * 60) + (13 * 60 * 60)
    assert_equal secs, @dl_time.to_i
  end


  def test_accessor_hours
    assert @dl_time.respond_to?(:hours)
    assert_equal 13, @dl_time.hours
  end


  def test_accessor_minutes
    assert @dl_time.respond_to?(:minutes)
    assert_equal 37, @dl_time.minutes
  end


  def test_accessor_seconds
    assert @dl_time.respond_to?(:seconds)
    assert_equal 42, @dl_time.seconds
  end


  def test_alias_hour
    assert @dl_time.respond_to?(:hour)
    assert_equal 13, @dl_time.hour
  end


  def test_alias_min
    assert @dl_time.respond_to?(:min)
    assert_equal 37, @dl_time.min
  end


  def test_alias_sec
    assert @dl_time.respond_to?(:sec)
    assert_equal 42, @dl_time.sec
  end



  def test_strftime
    assert_equal "13:37", @dl_time.strftime("%H:%M")
    assert_equal "13:37:42", @dl_time.strftime("%H:%M:%S")
    assert_equal "13-37-42", @dl_time.strftime("%H-%M-%S")

    assert_equal "1:37 pm", @dl_time.strftime("%-l:%M %P")
  end


end


test/test_helper.rb
require "dateless_time"

require 'minitest/autorun'
require 'minitest/pride'


test/creation_test.rb
require 'test_helper'

class CreationTest < Minitest::Test


  def test_create_with_now
    @time = Time.now
    @dl_time = DatelessTime.now

    assert_equal @time.hour, @dl_time.hours
    assert_equal @time.min, @dl_time.minutes
    assert_equal @time.sec, @dl_time.seconds
    assert_equal @time.strftime("%H:%M:%S"), @dl_time.to_s
    assert_equal (@time.sec + (@time.min * 60) + (@time.hour * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_time
    @time = Time.new 2014, 1, 1, 13, 37, 42
    @dl_time = DatelessTime.new @time

    assert_equal 13, @dl_time.hours
    assert_equal 37, @dl_time.minutes
    assert_equal 42, @dl_time.seconds
    assert_equal (42 + (37 * 60) + (13 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_datetime
    @datetime = DateTime.new 2014, 1, 1, 13, 37, 42
    @dl_time = DatelessTime.new @datetime

    assert_equal 13, @dl_time.hours
    assert_equal 37, @dl_time.minutes
    assert_equal 42, @dl_time.seconds
    assert_equal (42 + (37 * 60) + (13 * 60 * 60)), @dl_time.to_i
  end



  def test_create_with_string
    @dl_time = DatelessTime.new "12:13:14"
  
    assert_equal 12, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 14, @dl_time.seconds
    assert_equal (14 + (13 * 60) + (12 * 60 * 60)), @dl_time.to_i
  end



  def test_create_with_string_without_seconds
    @dl_time = DatelessTime.new "12:13"
  
    assert_equal 12, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal ((13 * 60) + (12 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_string_without_seconds_and_minutes
    @dl_time = DatelessTime.new "12"
  
    assert_equal 12, @dl_time.hours
    assert_equal 0, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal (12 * 60 * 60), @dl_time.to_i
  end



  def test_create_with_bad_string
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new "hello"
    end
  end


  def test_create_with_bad_numerical_string
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new "1000"
    end
  end


  def test_create_with_empty_string
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new("")
    end
  end


  def test_create_with_string_with_am
    @dl_time = DatelessTime.new "10:13:14 am"
  
    assert_equal 10, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 14, @dl_time.seconds
    assert_equal (14 + (13 * 60) + (10 * 60 * 60)), @dl_time.to_i

    @dl_time = DatelessTime.new "10:13AM"
  
    assert_equal 10, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal ((13 * 60) + (10 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_string_with_pm
    @dl_time = DatelessTime.new "10:13:14 pm"
  
    assert_equal 22, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 14, @dl_time.seconds
    assert_equal (14 + (13 * 60) + (22 * 60 * 60)), @dl_time.to_i

    @dl_time = DatelessTime.new "10:13PM"
  
    assert_equal 22, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal ((13 * 60) + (22 * 60 * 60)), @dl_time.to_i
  end



  def test_create_with_string_with_nonsensical_pm
    assert_raises DatelessTime::TimeOutOfRangeError do
      @dl_time = DatelessTime.new "15:13:14 pm"
    end
  
    assert_raises DatelessTime::TimeOutOfRangeError do
      @dl_time = DatelessTime.new "13:13PM"
    end
  end




  def test_create_with_borderline_am
    @dl_time = DatelessTime.new "12:13:14 am"
    assert_equal 0, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 14, @dl_time.seconds
    assert_equal (14 + (13 * 60) + 0), @dl_time.to_i
  end


  def test_create_with_borderline_pm
    @dl_time = DatelessTime.new "12:13:14 pm"
    assert_equal 12, @dl_time.hours
    assert_equal 13, @dl_time.minutes
    assert_equal 14, @dl_time.seconds
    assert_equal (14 + (13 * 60) + (12 * 60 * 60)), @dl_time.to_i
  end

  

  def test_create_with_seconds
    secs = 59 + (35 * 60) + (8 * 60 * 60)
    @dl_time = DatelessTime.new secs

    assert_equal 8, @dl_time.hours
    assert_equal 35, @dl_time.minutes
    assert_equal 59, @dl_time.seconds
    assert_equal secs, @dl_time.to_i
  end


  def test_create_with_very_few_seconds
    secs = 10 + (2 * 60)
    @dl_time = DatelessTime.new secs

    assert_equal 0, @dl_time.hours
    assert_equal 2, @dl_time.minutes
    assert_equal 10, @dl_time.seconds
    assert_equal secs, @dl_time.to_i
  end


  def test_create_with_zero_seconds
    @dl_time = DatelessTime.new 0

    assert_equal 0, @dl_time.hours
    assert_equal 0, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal 0, @dl_time.to_i
  end


  def test_create_with_too_many_seconds
    assert_raises DatelessTime::TimeOutOfRangeError do
      @dl_time = DatelessTime.new(DatelessTime::SECONDS_IN_24_HOURS + 1)
    end
  end



  def test_create_with_hash
    @dl_time = DatelessTime.new({ hours: 1, minutes: 30, seconds: 45 })

    assert_equal 1, @dl_time.hours
    assert_equal 30, @dl_time.minutes
    assert_equal 45, @dl_time.seconds
    assert_equal (45 + (30 * 60) + (1 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_hash_with_extra_keys
    @dl_time = DatelessTime.new({ hours: 1, minutes: 30, seconds: 45, foo: "bar" })

    assert_equal 1, @dl_time.hours
    assert_equal 30, @dl_time.minutes
    assert_equal 45, @dl_time.seconds
    assert_equal (45 + (30 * 60) + (1 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_hash_without_seconds
    @dl_time = DatelessTime.new({ hours: 1, minutes: 30 })

    assert_equal 1, @dl_time.hours
    assert_equal 30, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal ((30 * 60) + (1 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_hash_without_hours
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new({ minutes: 30 })
    end
  end


  def test_create_with_empty_hash
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new({})
    end
  end


  def test_create_with_hash_without_hours_but_other_random_keys
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new({ batman: "nananana" })
    end
  end


  def test_create_with_hash_without_seconds_and_minutes
    @dl_time = DatelessTime.new({ hours: 1 })

    assert_equal 1, @dl_time.hours
    assert_equal 0, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal (1 * 60 * 60), @dl_time.to_i
  end


  def test_create_with_array
    @dl_time = DatelessTime.new [2, 11, 44]

    assert_equal 2, @dl_time.hours
    assert_equal 11, @dl_time.minutes
    assert_equal 44, @dl_time.seconds
    assert_equal (44 + (11 * 60) + (2 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_array_with_extra_elements
    @dl_time = DatelessTime.new [2, 11, 44, 1, 2, 3]

    assert_equal 2, @dl_time.hours
    assert_equal 11, @dl_time.minutes
    assert_equal 44, @dl_time.seconds
    assert_equal (44 + (11 * 60) + (2 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_array_without_seconds
    @dl_time = DatelessTime.new [2, 11]

    assert_equal 2, @dl_time.hours
    assert_equal 11, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal ((11 * 60) + (2 * 60 * 60)), @dl_time.to_i
  end


  def test_create_with_array_without_seconds_and_minutes
    @dl_time = DatelessTime.new [2]

    assert_equal 2, @dl_time.hours
    assert_equal 0, @dl_time.minutes
    assert_equal 0, @dl_time.seconds
    assert_equal (2 * 60 * 60), @dl_time.to_i
  end


  def test_create_with_empty_array
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new([])
    end
  end


  def test_create_with_array_with_bad_data
    assert_raises DatelessTime::TimeOutOfRangeError do
      @dl_time = DatelessTime.new([25, 10, 11])
    end

    assert_raises DatelessTime::TimeOutOfRangeError do
      @dl_time = DatelessTime.new([11, 67, 11])
    end

    assert_raises DatelessTime::TimeOutOfRangeError do
      @dl_time = DatelessTime.new([9, 10, 100])
    end
  end


  def test_create_with_other_dateless_time
    @other = DatelessTime.now
    @this  = DatelessTime.new(@other)

    assert_equal @other.hours, @this.hours
    assert_equal @other.minutes, @this.minutes
    assert_equal @other.seconds, @this.seconds
    assert_equal @other.to_i, @this.to_i
  end



  def test_create_without_argument
    @time = Time.now
    @dl_time = DatelessTime.new

    assert_equal @time.hour, @dl_time.hours
    assert_equal @time.min, @dl_time.minutes
    assert_equal @time.sec, @dl_time.seconds
    assert_equal @time.strftime("%H:%M:%S"), @dl_time.to_s
    assert_equal (@time.sec + (@time.min * 60) + (@time.hour * 60 * 60)), @dl_time.to_i
  end



  def test_create_with_nil
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new nil
    end
  end


  def test_create_with_unsupported_objects
    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new (1..4)
    end

    assert_raises DatelessTime::InitializationError do
      @dl_time = DatelessTime.new 1.21
    end
  end
end


lib/static_time.rb
require "static_time/version"

class StaticTime

  SECONDS_IN_24_HOURS = 86400

  MAX_HOURS = 24
  MAX_MINUTES = 59
  MAX_SECONDS = 59
  
  TIME_STRING_REGEX = /\A\d{1,2}(:\d{2})?(:\d{2})?( ?(am|pm))?\z/i
  AM_PM_REGEX = /( )?(am|pm)\z/i

  SPRINTF_FORMAT = "%02d:%02d:%02d".freeze



  attr_reader :hours, :minutes, :seconds
  

  def self.now
    new
  end


  def initialize(source = Time.now)
    conditional_init source
  end



  def to_time(base = Time.now)
    @time_value ||= Time.new(base.year, base.month, base.day,
                        @hours, @minutes, @seconds, base.utc_offset)
  rescue
    nil
  end


  def to_s
    @string_value ||= sprintf(SPRINTF_FORMAT, @hours, @minutes, @seconds)
  rescue
    nil
  end


  def to_h
    @hash_value ||= { hours: @hours, minutes: @minutes, seconds: @seconds }
  end


  def to_a
    @array_value ||= [@hours, @minutes, @seconds]
  end


  def seconds_since_midnight
    @seconds_since_midnight ||= calculate_seconds_since_midnight
  end

  alias_method :to_i, :seconds_since_midnight


  def strftime(template)
    to_time.strftime(template)
  rescue
    nil
  end



private


  def conditional_init(source)
    case source
    when Time   then init_with_time(source)
    when String then init_with_string(source)
    when Fixnum then init_with_seconds(source)
    when Hash   then init_with_hash(source)
    when Array  then init_with_array(source)
    else raise StaticTime::InitializationError
    end
  end


  def init_with_time(time)
    @hours   = time.hour
    @minutes = time.min
    @seconds = time.sec
    seconds_since_midnight
  end


  def init_with_string(string)
    validate_time_string string
    data = time_string_to_array string
    init_with_array data
  end


  def init_with_seconds(seconds)
    validate_seconds_since_midnight seconds
    #seconds = SECONDS_IN_24_HOURS if seconds > SECONDS_IN_24_HOURS
    @seconds_since_midnight = seconds
    calculate_hours_minutes_and_seconds
  end


  def init_with_hash(hash)
    validate_time_hash hash
    @hours   = hash[:hours]
    @minutes = hash[:minutes] || 0
    @seconds = hash[:seconds] || 0
    seconds_since_midnight
  end


  def init_with_array(array)
    validate_time_array array
    @hours   = array[0]
    @minutes = array[1] || 0
    @seconds = array[2] || 0
    seconds_since_midnight
  end





  def calculate_seconds_since_midnight
    if @hours && @minutes && @seconds
      cache =  @seconds
      cache += @minutes * 60
      cache += @hours * 3600
      validate_seconds_since_midnight cache
      cache
    else
      nil
    end
  end


  def calculate_hours_minutes_and_seconds
    if @seconds_since_midnight
      cache, @seconds = @seconds_since_midnight.divmod(60)
      @hours, @minutes = cache.divmod(60)
    else
      nil
    end
  end


  def validate_time_string(str)
    unless TIME_STRING_REGEX =~ str
      raise StaticTime::InitializationError, "bad string format"
    end
  end


  # COLON = ":".freeze
  # EMPTY = "".freeze
  # PM = "pm".freeze

  def time_string_to_array(str)
    am_pm = false

    if AM_PM_REGEX =~ str
      am_pm = str[-2,2].downcase
      str.sub!(AM_PM_REGEX, '')
    end

    ary = str.split(":").map(&:to_i)
    ary[0] += 12 if (am_pm && am_pm == 'pm')
    ary
  end


  def validate_time_array(ary)
    if ary.empty?
      raise StaticTime::InitializationError
    elsif ary[0] > MAX_HOURS || (ary[1] && ary[1] > MAX_MINUTES) || (ary[2] && ary[2] > MAX_SECONDS)
      raise StaticTime::TimeOutOfRangeError
    end
  end


  def validate_time_hash(h)
    if h.empty? || h[:hours].nil?
      raise StaticTime::InitializationError
    elsif h[:hours] > MAX_HOURS || (h[:minutes] && h[:minutes] > MAX_MINUTES) || (h[:seconds] && h[:seconds] > MAX_SECONDS)
      raise StaticTime::TimeOutOfRangeError
    end
  end


  def validate_seconds_since_midnight(seconds)
    if seconds > SECONDS_IN_24_HOURS
      raise StaticTime::TimeOutOfRangeError
    end
  end

end



class StaticTime::InitializationError < StandardError
end

class StaticTime::TimeOutOfRangeError < StaticTime::InitializationError
end


lib/dateless_time.rb
require 'date'
require 'dateless_time/version'

class DatelessTime

  include Comparable

  SECONDS_IN_24_HOURS = 86400

  MAX_HOURS = 24
  MAX_MINUTES = 59
  MAX_SECONDS = 59
  
  TIME_STRING_REGEX = /\A\d{1,2}(:\d{2})?(:\d{2})?( ?(am|pm))?\z/i
  AM_PM_REGEX = /( )?(am|pm)\z/i

  SPRINTF_FORMAT = "%02d:%02d:%02d".freeze



  attr_reader :hours, :minutes, :seconds
  alias hour hours
  alias min minutes
  alias sec seconds

  def self.now
    new
  end


  def initialize(source = Time.now)
    conditional_init source
  end



  def to_time(base = Time.now)
    args = [base.year, base.month, base.day, @hours, @minutes, @seconds]

    if base.is_a?(Time)
      args << base.utc_offset
    elsif base.is_a?(DateTime)
      args << (base.offset * SECONDS_IN_24_HOURS)
    end

    Time.new(*args)
  rescue
    nil
  end


  def to_datetime(base = DateTime.now)
    args = [base.year, base.month, base.day, @hours, @minutes, @seconds]

    normalized_offset = begin
      if base.is_a?(DateTime)
        base.offset * 24
      elsif base.is_a?(Time)
        base.utc_offset / 3600
      end
    end

    normalized_offset && args << sprintf("%+d", normalized_offset)

    DateTime.new(*args)
  rescue
    nil
  end


  def to_s
    @string_value ||= sprintf(SPRINTF_FORMAT, @hours, @minutes, @seconds)
  rescue
    nil
  end


  def to_h
    @hash_value ||= { hours: @hours, minutes: @minutes, seconds: @seconds }
  end


  def to_a
    @array_value ||= [@hours, @minutes, @seconds]
  end


  def seconds_since_midnight
    @seconds_since_midnight ||= calculate_seconds_since_midnight
  end

  alias_method :to_i, :seconds_since_midnight


  def strftime(template)
    to_time.strftime(template)
  rescue
    nil
  end


  def <=>(other)
    raise TypeError unless other.is_a?(DatelessTime)
    to_i <=> other.to_i
  end



private


  def conditional_init(source)
    case source
    when Time, DateTime, DatelessTime then init_with_time(source)
    when String then init_with_string(source)
    when Fixnum then init_with_seconds(source)
    when Hash   then init_with_hash(source)
    when Array  then init_with_array(source)
    else raise DatelessTime::InitializationError, "DatelessTime objects cannot be initialized with instances of #{source.class}."
    end
  end


  def init_with_time(time)
    @hours   = time.hour
    @minutes = time.min
    @seconds = time.sec
    seconds_since_midnight
  end



  def init_with_string(string)
    validate_time_string string
    data = time_string_to_array string
    init_with_array data
  end


  def init_with_seconds(seconds)
    validate_seconds_since_midnight seconds
    #seconds = SECONDS_IN_24_HOURS if seconds > SECONDS_IN_24_HOURS
    @seconds_since_midnight = seconds
    calculate_hours_minutes_and_seconds
  end


  def init_with_hash(hash)
    validate_time_hash hash
    @hours   = hash[:hours]
    @minutes = hash[:minutes] || 0
    @seconds = hash[:seconds] || 0
    seconds_since_midnight
  end


  def init_with_array(array)
    validate_time_array array
    @hours   = array[0]
    @minutes = array[1] || 0
    @seconds = array[2] || 0
    seconds_since_midnight
  end



  def calculate_seconds_since_midnight
    if @hours && @minutes && @seconds
      cache =  @seconds
      cache += @minutes * 60
      cache += @hours * 3600
      validate_seconds_since_midnight cache
      cache
    else
      nil
    end
  end


  def calculate_hours_minutes_and_seconds
    if @seconds_since_midnight
      cache, @seconds = @seconds_since_midnight.divmod(60)
      @hours, @minutes = cache.divmod(60)
    else
      nil
    end
  end


  def validate_time_string(str)
    unless TIME_STRING_REGEX =~ str
      raise DatelessTime::InitializationError, "bad string format"
    end
  end


  # COLON = ":".freeze
  # EMPTY = "".freeze
  # PM = "pm".freeze

  def time_string_to_array(str)
    am_pm = false

    if AM_PM_REGEX =~ str
      am_pm = str[-2,2].downcase
      str.sub!(AM_PM_REGEX, '')
    end

    ary = str.split(":").map(&:to_i)
    ary[0] = adjust_for_am_pm(ary[0], am_pm) if am_pm
    ary
  end


  def adjust_for_am_pm(hour, am_pm)
    if am_pm == 'am'
      if hour == 12
        hour = 0
      end
    elsif am_pm == 'pm'
      if hour != 12
        hour += 12
      end
    end
    hour
  end



  def validate_time_array(ary)
    if ary.empty?
      raise DatelessTime::InitializationError
    elsif ary[0] > MAX_HOURS || (ary[1] && ary[1] > MAX_MINUTES) || (ary[2] && ary[2] > MAX_SECONDS)
      raise DatelessTime::TimeOutOfRangeError
    end
  end


  def validate_time_hash(h)
    if h.empty? || h[:hours].nil?
      raise DatelessTime::InitializationError
    elsif h[:hours] > MAX_HOURS || (h[:minutes] && h[:minutes] > MAX_MINUTES) || (h[:seconds] && h[:seconds] > MAX_SECONDS)
      raise DatelessTime::TimeOutOfRangeError
    end
  end


  def validate_seconds_since_midnight(seconds)
    if seconds > SECONDS_IN_24_HOURS
      raise DatelessTime::TimeOutOfRangeError
    end
  end

end



class DatelessTime::InitializationError < StandardError
end

class DatelessTime::TimeOutOfRangeError < DatelessTime::InitializationError
end
